numpy/core/numeric.py

```diff
@@ -2379,6 +2379,9 @@ def within_tol(x, y, atol, rtol):
 def _array_equal_dispatcher(a1, a2, equal_nan=None):
     return (a1, a2)
 
+def _dtype_cannot_hold_nan(dtype):
+    return dtype in {np.int8, np.int16, np.int32, np.int64}
```

Comment:
```suggestion
def _dtype_cannot_hold_nan(dtype):
    return dtype.type in {np.int8, np.int16, np.int32, np.int64}
```

Reply:
Merging both your suggestions:

```suggestion
_no_nan_dtypes = {np.int8, np.int16, np.int32, np.int64}

def _dtype_cannot_hold_nan(dtype):
    return dtype.type in _no_nan_dtypes
```